source/common/buffer/buffer_impl.h

```diff
@@ -297,13 +303,44 @@ class Slice {
     return num_pages * PageSize;
   }
 
+  static StoragePtr newStorage(uint64_t capacity) {
+    ASSERT(sliceSize(default_slice_size_) == default_slice_size_,
+           "default_slice_size_ incompatible with sliceSize()");
+
+    StoragePtr storage;
+    if (capacity == default_slice_size_ && !free_list_.empty()) {
+      storage = std::move(free_list_.back());
+      free_list_.pop_back();
+    } else {
+      storage.reset(new uint8_t[capacity]);
+    }
+
+    return storage;
+  }
+
+  static void freeStorage(StoragePtr storage, uint64_t capacity) {
+    if (storage == nullptr) {
+      return;
+    }
+
+    if (capacity == default_slice_size_ && free_list_.size() < free_list_max_) {
+      free_list_.emplace_back(std::move(storage));
+      ASSERT(storage == nullptr);
+    } else {
+      storage.reset();
+    }
+  }
+
+  static constexpr uint32_t free_list_max_ = 8;
+  static thread_local absl::InlinedVector<StoragePtr, free_list_max_> free_list_;
```

Comment:
Worth disabling this freelist in memory constrained environments (e.i. Envoy mobile)

Reply:
I don't know that it will have much of an effect. That much peak memory will need to be available for every read operation. Reading a socket is such a frequent operation that it probably won't help. Also, given that large allocations are often directly backed by mmap, there's a good chance the page won't be allocated by the kernel until it is written to, which shouldn't happen in the case where it's put immediately back into the freelist.